cEP-0031.md

```diff
@@ -0,0 +1,254 @@
+
+```py
+class PipRequirement(PackageRequirement):
+    def get_version_info(self):
+        # This would get the package info in json format
+        url = 'https://pypi.python.org/pypi/%s/json' % (self.package,)
+
+        # Get the required data
+        data = json.load()
+        versions = data['releases'].keys().sort(key = StrictVersion)
+        return versions
+```
+
+- For npm dependencies, we can use the npm cli command `npm outdated` which
+    produces an output as below:
+
+```
+$ npm outdated
+Package      Current   Wanted   Latest  Location
+glob          5.0.15   5.0.15    6.0.1  test-outdated-output
+nothingness    0.0.3      git      git  test-outdated-output
+npm            3.5.1    3.5.2    3.5.1  test-outdated-output
+local-dev      0.0.3   linked   linked  test-outdated-output
+once           1.3.2    1.3.3    1.3.3  test-outdated-output
+```
+
+The output can be directly parsed and can be used to get the list of all
+outdated dependencies.
+
+### 2. FileExistBear
+
+Issue: [coala/coala-bears#527][527]
+
+The bear will be responsible to check if the files exist in the project
+directory. The name of the files can be passed as a regex to the bear, to
+which the bear will check their existence.
+
+```py
+class FileExistBear(LocalBear):
+    def run(self, required_files: tuple,):
+        for file in required_files:
+            if not os.path.is_file():
+                message = 'File {} not found.'.format(file)
+                yield Result.from_values(origin=self, message=message,)
+```
+
+### 3. RegexLintBear
+
+Issue: [coala/coala-bears#1532][1532]
+
+The task of the bear is to check the regex in strings. This can be done using
+the AnnotationBear to detect all the regex strings and then check for the valid
+regex through the prepared algorithm for each type.
+
+For Python, the bear can be written as follows:
+
+```py
+def is_regex(code, language):
+    """
+    Returns true if the string is a regex.
+    """
+
+class RegexLintBear(LocalBear):
+    def run(self, file, filename, language,):
+        strings = run_local_bear(AnnotationBear)
+        for string in strings:
+            if is_regex(string, language):
+                try:
+                    re.compile(string)
+                except re.error as Exception:
+                    message = 'This is not a valid regex.'
+                    yield Result.from_values(origin=self, message=message)
+```
+
+### 4. FileModeBear
+
+Issue: [coala/coala-bears#2370][2370]
+
+The bear will check the permissions on the files provided by the user
+and to ensure that the file permissions are the one that is expected.
+
+The bear will be used as follows:
+
+```
+[all.mode]
+bears = FileModeBear
+filemode = rw
+
+[all.shell.mode]
+bears = FileModeBear
+filemode = rwx
+```
+
+The bear will first find the permissions of the files specified by the user and
+then if the permissions are not suitable, the bear will try to change those
```

Comment:
Bears are not allowed to make modifications -- they need to create and return `Action` objection, which the user chooses to apply or not apply.  If you want to make changes to file permissions, this will involve enhancing core to know file permissions and have Actions to modify file permissions.

This needs to be spec'd out clearly in the cEP , and that will be difficult, so I suggest you drop this modification thing from the first iteration of the cEP , and try to add it in a followup PR

Reply:
Yep, I forgot to update this. The modifications are not a part of the bear right now.